ENH: Worker Environments

TomAugspurger opened the pull request:
Adds the notion of worker environments to the client and scheduler.
These allow for specifying certain setup and teardown methods that
should be called when workers enter or exit the pool. In the future
environments could allow for scheduling decisions, e.g. restricting
certain tasks to certain environments.

```python
>>> client.register_worker_environment("my-env",
...                                    condition=lambda: True)
```

This is pretty rough at the moment.

mrocklin commented on distributed/tests/test_worker_environments.py:
```diff
@@ -0,0 +1,96 @@
+    start = time()
+    while not w.status == 'closed':
+        yield gen.sleep(0.01)
+        assert time() - start < 5
```
Is this timing block necessary?

mrocklin commented on distributed/tests/test_worker_environments.py:
```diff
@@ -0,0 +1,96 @@
+def test_register_condition_for_worker_raises(c, s, a, b):
+    r = c._register_worker_environment('myenv', lambda: True)
+    with pytest.raises(TypeError) as m:
+        r.result()
```
FWIW I *always* use the `yield` syntax everywhere throughout the project.

mrocklin commented on distributed/worker.py:
```diff
@@ -2233,6 +2238,47 @@ def _get_client(self, timeout=3):
+            if name not in self.environments:
+                try:
+                    env = pickle.loads(env)
+                    meets_condition = yield self.executor_submit('condition', env.condition)
```
```python
from .comm.utils import offload
meets_condition = yield offload(env.condition)
```

We should move this to `.utils`

mrocklin commented on distributed/worker.py:
```diff
@@ -2233,6 +2238,47 @@ def _get_client(self, timeout=3):
+                    # TODO: Should this be in some kind of "while running" loop with
+                    # a timeout?
+                    if gen.is_future(meets_condition):
+                        meets_condition = meets_condition.result()
```
Prefer `yield` if that's ok

mrocklin commented on distributed/worker.py:
```diff
@@ -2233,6 +2238,47 @@ def _get_client(self, timeout=3):
+                    if meets_condition:
+                        # TODO: This currently blocks. I expect that that is
+                        # not desirable, as a long setup task may look like a
+                        # worker that's become unresponsive.
```
This doesn't seem to block anymore.  I suspect that this comment is out of date.

mrocklin commented on distributed/worker.py:
```diff
@@ -2233,6 +2238,47 @@ def _get_client(self, timeout=3):
+                        # TODO: This currently blocks. I expect that that is
+                        # not desirable, as a long setup task may look like a
+                        # worker that's become unresponsive.
+                        yield self.executor_submit('setup', env.setup)
```
What if setup/condition are coroutines?  This might not function well.

Actually, I'm tempted to just run them locally as is without a separate thread.  If people care about non-blocking behavior then they should write coroutines.  I think that we can expect a bit of expertise from anyone using Environments.  Most of the time this probably won't be an issue.

TomAugspurger commented:
> `starting all methods with environment_` 

Do you prefer `environment_register` or `worker_environment_register`? Do you anticipate any non-worker environments in the future?

mrocklin commented:
environment_register sounds good to me

On Fri, Aug 25, 2017 at 1:33 PM, Tom Augspurger <notifications@github.com>
wrote:

> starting all methods with environment_
>
> Do you prefer environment_register or worker_environment_register? Do you
> anticipate any non-worker environments in the future?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/distributed/pull/1358#issuecomment-324987423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASszICg7OgiNq58XLhm0Y9PyjeHS_XDks5sbwVngaJpZM4PB1W7>
> .
>

TomAugspurger commented:
Just to make sure I understand the running of methods on workers correctly:

> Actually, I'm tempted to just run them locally as is without a separate thread. If people care about non-blocking behavior then they should write coroutines. I think that we can expect a bit of expertise from anyone using Environments. Most of the time this probably won't be an issue.

What would that look like on `Worker.environment_register`? Since distributed is calling the function or coroutine, aren't we responsible for calling it on the main thread or not? Perhaps I misunderstood.

---

My latest push simplifies things:

- Renamed all `register_worker_environmnet` style methods to `environment_register` style
- Removed essentially all exception handling, letting `.rpc` handle it for us
- Added environment de-registration
- Make all methods (de)register a single environment at once (this was the case for the `Client` already). There's a small utility method on the scheduler for registering many environments, when a new worker is added
- Made `WorkerEnvironment` methods regular functions (not coroutines).
- Call `WorkerEnvironment` methods on the main thread

I'm not sure if those last two are desirable or not.

TomAugspurger commented on distributed/environments.py:
```diff
@@ -0,0 +1,52 @@
+        in the environment.
+        """
+        if self._condition:
+            return self._condition()
```
FYI, `gen.maybe_future` seems to be deprecated: http://www.tornadoweb.org/en/stable/gen.html#tornado.gen.maybe_future.

TomAugspurger commented:
Moving [this](https://github.com/dask/distributed/pull/1358#discussion_r135114979) comment up, since I haven't addressed it yet.

# Summary of Changes

Equality Semantics

> This will happen concurrently with other things. We might want to avoid either having the worker start work or having it even be known in the scheduler until the environment membership is known.
> 
> I see two options:
> 
> 1. Change add_worker and remove_worker to be coroutines. Use `self.rpc(address).register_environment(...)` before we kick things off. This is best because the worker won't be assigned any work until this comes back.
> 2. Add environments to the handshake that occurs in the beginning of handle_worker (look for 'compute-stream').
> This could be part of the original handshake or could come directly afterwards or before. This is simpler, doesn't give the worker any work before its environment suite is known, but might inadvertently schedule work beforehand.
>
> Actually, if all scheduling is going to be environment inclusive (e.g. run this task on a GPU node) then the second case might suffice.

mrocklin commented on docs/source/worker-environments.rst:
```diff
@@ -0,0 +1,84 @@
+   has a dictionary ``environment_state :: name -> Dict``, where you can get and
+   put whatever you want in that nested ``Dict``)
+
+TODO: Can / should the setup or teardown methods take the ``dask_worker`` itself?
```
This sounds useful to me.

mrocklin replied:
In `client.run` we magically look for the special `dask_worker` argument, and if it is there then we populate it accordingly.  I'm not sure if we want to repeat this pattern, but it has so far worked out well.

mrocklin commented on docs/source/worker-environments.rst:
```diff
@@ -0,0 +1,84 @@
+   leave the environment
+5. A ``state`` variable (TODO: what exactly is this? Tentatively, each worker
+   has a dictionary ``environment_state :: name -> Dict``, where you can get and
+   put whatever you want in that nested ``Dict``)
```
Is having state necessary?  Perhaps drop this but include examples of environment instances with state?

mrocklin commented on docs/source/worker-environments.rst:
```diff
@@ -0,0 +1,84 @@
+.. code-block:: python
+
+   client.environment_register(name='high-memory',
+                               environment=HighMemory)
```
This example needs a class definition?

mrocklin commented:
This PR has gone stale.  Closing for now.
